src/services/completions.ts

```diff
@@ -1280,31 +1331,64 @@ namespace ts.Completions {
+                    // If `!!d.name.originalKeywordKind`, this is `export { _break as break };` -- skip this and prefer the keyword completion.
+                    if (some(symbol.declarations, d => isExportSpecifier(d) && !!d.propertyName && isIdentifierANonContextualKeyword(d.name))) {
+                        continue;
                     }
-
-                    const origin: SymbolOriginInfoExport = { kind: SymbolOriginInfoKind.Export, moduleSymbol, isDefaultExport };
-                    if (detailsEntryId || stringContainsCharactersInOrder(getSymbolName(symbol, origin, target).toLowerCase(), tokenTextLowerCase)) {
-                        symbols.push(symbol);
-                        symbolToSortTextMap[getSymbolId(symbol)] = SortText.AutoImportSuggestions;
-                        symbolToOriginInfoMap[getSymbolId(symbol)] = origin;
+                    // If `!!d.parent.parent.moduleSpecifier`, this is `export { foo } from "foo"` re-export, which creates a new symbol (thus isn't caught by the first check).
+                    if (some(symbol.declarations, d => isExportSpecifier(d) && !d.propertyName && !!d.parent.parent.moduleSpecifier)) {
+                        // Walk the export chain back one module (step 1 or 2 in diagrammed example)
+                        const nearestExportSymbol = Debug.assertDefined(getNearestExportSymbol(symbol));
+                        if (!symbolHasBeenSeen(nearestExportSymbol)) {
+                            aliasesToReturnIfOriginalsAreMissing.set(getSymbolId(nearestExportSymbol).toString(), { alias: symbol, moduleSymbol, insertAt: symbols.length });
+                            aliasesToAlreadyIncludedSymbols.set(getSymbolId(symbol).toString(), true);
+                        }
+                        else {
+                            // Perf - we know this symbol is an alias to one that’s already covered in `symbols`, so store it here
+                            // in case another symbol re-exports this one; that way we can short-circuit as soon as we see this symbol id.
+                            addToSeen(aliasesToAlreadyIncludedSymbols, getSymbolId(symbol));
+                        }
                     }
+                    else {
+                        // This is not a re-export, so see if we have any aliases pending and remove them (step 3 in diagrammed example)
+                        aliasesToReturnIfOriginalsAreMissing.delete(getSymbolId(symbol).toString());
+                        pushSymbol(symbol, moduleSymbol);
+                    }
+                }
+            });
+
+            // By this point, any potential duplicates that were actually duplicates have been
+            // removed, so the rest need to be added. (Step 4 in diagrammed example)
+            aliasesToReturnIfOriginalsAreMissing.forEach(({ alias, moduleSymbol, insertAt }) => pushSymbol(alias, moduleSymbol, insertAt));

```

Comment:
why is the order important? Seems like you want to insert at the index of the length of symbols at the time the backup symbol was created instead of the current length. But why?

Reply:
I’m pretty sure it’s not important, since completions get sorted based on other criteria later, but at the time I wasn’t yet confident of that and wanted to make the minimum possible change. But, I suppose there’s a small performance penalty—I’ll change to push and see if tests still pass (I think they will).